irohad/ametsuchi/impl/rocksdb_common.hpp

```diff
@@ -0,0 +1,1218 @@
+    if (status.IsNotFound())
+      throw IrohaDbError(
+          14,
+          fmt::format("{}. Was not found.", std::forward<F>(op_formatter)()));
+
+    if (!status.ok())
+      throw IrohaDbError(15,
+                         fmt::format("{}. Failed with status: {}.",
+                                     std::forward<F>(op_formatter)(),
+                                     status.ToString()));
+  }
+
+  template <typename F>
+  inline void canExist(rocksdb::Status status, F &&op_formatter) {
+    if (status.IsNotFound() || status.ok())
+      return;
+
+    throw IrohaDbError(18,
+                       fmt::format("{}. Failed with status: {}.",
+                                   std::forward<F>(op_formatter)(),
+                                   status.ToString()));
+  }
+
+  template <kDbEntry kSc, typename F>
+  static void checkStatus(rocksdb::Status status, F &&op_formatter) {
+    if constexpr (kSc == kDbEntry::kAll || kSc == kDbEntry::kMustExist)
+      return mustExist(status, std::forward<F>(op_formatter));
+    else if constexpr (kSc == kDbEntry::kMustNotExist)
+      return mustNotExist(status, std::forward<F>(op_formatter));
+    else if constexpr (kSc == kDbEntry::kCanExist)
+      return canExist(status, std::forward<F>(op_formatter));
+    else
+      assert(!"Unexpected status check value");
+  }
+
+  template <kDbOperation kOp, kDbEntry kSc, typename F, typename... Args>
+  rocksdb::Status executeOperation(RocksDbCommon &common,
+                                   F &&op_formatter,
+                                   Args &&... args) {
+    rocksdb::Status status;
+    if constexpr (kOp == kDbOperation::kGet || kOp == kDbOperation::kCheck)
+      status = common.get(std::forward<Args>(args)...);
+    else if constexpr (kOp == kDbOperation::kPut)
+      status = common.put(std::forward<Args>(args)...);
+    else if constexpr (kOp == kDbOperation::kDel)
+      status = common.del(std::forward<Args>(args)...);
+    else
+      assert(!"Unexpected operation value!");
+
+    checkStatus<kSc>(status, std::forward<F>(op_formatter));
+    return status;
+  }
+
+  /**
+   * Access to account quorum file.
+   * @tparam kOp @see kDbOperation
+   * @tparam kSc @see kDbEntry
+   * @tparam F callback with operation result
+   * @param common @see RocksDbCommon
+   * @param domain id
+   * @param account name
+   * @param func callback with the result
+   * @return determined by the callback
+   */
+  template <kDbOperation kOp = kDbOperation::kGet,
+            kDbEntry kSc = kDbEntry::kAll,
+            typename F>
+  inline auto forQuorum(RocksDbCommon &common,
+                        std::string_view domain,
+                        std::string_view account,
+                        F &&func) {
+    assert(!domain.empty());
+    assert(!account.empty());
+
+    auto status = executeOperation<kOp, kSc>(
+        common,
+        [&] { return fmt::format("Account {}#{}", account, domain); },
```

Comment:
Is this optimization? Call format only on error, and do not compose message if no error.?

Reply:
yep